### PR TITLE
*: test update for Go 1.12.5 and related changes

### PR DESCRIPTION
**Pull request by spzala**
Testing update for Go 1.12.5. Remove deprecated unused and gosimple
pacakges, and mask staticcheck 1006.

Related #10528 #10438

Co-Authored-By: Gyuho Lee <leegyuho@amazon.com>

### CODE REVIEW

**Comment by spzala** [comment]
@gyuho hi, there are several fmt errors related to staticchecks (existing (e.g. SA2002) and newly added (e.g. S1002)). Should we try fixing them under this PR? Thanks!

**Comment by gyuho** [comment]
@spzala Yes, please. We want to fix the CIs. Thanks.

**Comment by spzala** [comment]
@gyuho OK, sounds great, thanks! Working on it!

**Comment by spzala** [comment]
@gyuho @hexfusion `staticchecks` errors seems under control but now as you can see there is a `unconvert` failure in a generated file `keys.generated.go`. I have tried taking care of `unconvert` failure in few other places but any thoughts on how should I handle the generated code? Thanks!

**Comment by spzala** [comment]
> @gyuho @hexfusion `staticchecks` errors seems under control but now as you can see there is a `unconvert` failure in a generated file `keys.generated.go`. I have tried taking care of `unconvert` failure in few other places but any thoughts on how should I handle the generated code? Thanks!

We will get rid of `client/keys.generated.go` if we go with https://github.com/etcd-io/etcd/pull/10667

**Comment by spzala** [comment]
@gyuho @hexfusion the changes are now ready for your initial review. Thanks!

**Comment by xiang90** [comment]
@gyuho can you take a look?

File: clientv3/balancer/grpc1.7-health.go [hunk 1]
```diff
@@ -507,7 +507,7 @@ func (b *GRPC17Health) Get(ctx context.Context, opts grpc.BalancerGetOptions) (g
 		addr = b.pinAddr
 		b.mu.RUnlock()
 		if closed {
-			return grpc.Address{Addr: ""}, nil, grpc.ErrClientConnClosing
+			return grpc.Address{Addr: ""}, nil, context.Canceled
```
Comment (xiang90) [inline clientv3/balancer/grpc1.7-health.go]: why do we need to change this error type?

File: clientv3/options.go [hunk 2]
```diff
@@ -47,7 +47,7 @@ var (
 	// client-side streaming retry limit, only applied to requests where server responds with
 	// a error code clearly indicating it was unable to process the request such as codes.Unavailable.
 	// If set to 0, retry is disabled.
-	defaultStreamMaxRetries = uint(^uint(0)) // max uint
+	defaultStreamMaxRetries = uint(0) // max uint
```
Comment (xiang90) [inline clientv3/options.go]: this does not seem to be right. we still want all 1's in the binary representation.

File: clientv3/balancer/grpc1.7-health.go (hunk 1 above)
Comment (gyuho) [reply clientv3/balancer/grpc1.7-health.go]: @spzala 

```
use he status code of Canceled instead of grpc.ErrClientConnClosing which is deprecated
```

Is it deprecated in gRPC? We haven't bumped up the gRPC yet.

File: clientv3/options.go (hunk 2 above)
Comment (spzala) [reply clientv3/options.go]: My bad, I will fix it. Thanks @xiang90 for catching it.

File: clientv3/balancer/grpc1.7-health.go (hunk 1 above)
Comment (spzala) [reply clientv3/balancer/grpc1.7-health.go]: @gyuho thanks and yes, the deprecated one. I also found a related issue https://github.com/etcd-io/etcd/issues/9304 so seems like we are good for now and I will make updates here. Thanks both!

**Comment by spzala** [comment]
Due to the merge of https://github.com/etcd-io/etcd/pull/10640 few things are changed, rebased but running into few conflicts and working on it.

**Comment by spzala** [comment]
@xiang90 @gyuho since 'staticcheck' and 'unconvert' are removed from `test` under #10640 I believe many of the changes made in this PR to take care related fmt errors are not needed to pass CI but if the changes are harmless I guess it's good to still keep them as a good practice? I have addressed your review comments and also opened https://github.com/etcd-io/etcd/issues/10703 for the future work. Thanks!

**Comment by xiang90** [comment]
travis failed with "go: sigs.k8s.io/yaml@v1.1.0: unrecognized import path "sigs.k8s.io/yaml" (https fetch: Get https://sigs.k8s.io/yaml?go-get=1: net/http: TLS handshake timeout)".

it seems to have something to do with https://github.com/etcd-io/etcd/commit/42a7ea6d337a2fecc9162676407aa94e6a2b1f1e. /cc @rohitsardesai83

**Comment by spzala** [comment]
> travis failed with "go: sigs.k8s.io/yaml@v1.1.0: unrecognized import path "sigs.k8s.io/yaml" (https fetch: Get https://sigs.k8s.io/yaml?go-get=1: net/http: TLS handshake timeout)".
> 
> it seems to have something to do with [42a7ea6](https://github.com/etcd-io/etcd/commit/42a7ea6d337a2fecc9162676407aa94e6a2b1f1e). /cc @rohitsardesai83

Thanks @xiang90 and yes, the same failure I noticed with another PR that came in today https://github.com/etcd-io/etcd/pull/10704

**Comment by rohitsardesai83** [comment]
> travis failed with "go: sigs.k8s.io/yaml@v1.1.0: unrecognized import path "sigs.k8s.io/yaml" (https fetch: Get https://sigs.k8s.io/yaml?go-get=1: net/http: TLS handshake timeout)".
> 
> it seems to have something to do with [42a7ea6](https://github.com/etcd-io/etcd/commit/42a7ea6d337a2fecc9162676407aa94e6a2b1f1e). /cc @rohitsardesai83

Hi @xiang90 , I cloned a fresh copy of etcd repo and tested with go 1.12.4

```sh
root@ubuntu-20:~/go/src/go.etcd.io/etcd# ./test
Running with TEST_CPUS: 1,2,4
Starting 'fmt' pass at Sat May  4 11:21:46 IST 2019
'shellcheck' started at Sat May  4 11:21:46 IST 2019
'shellcheck' completed at Sat May  4 11:21:46 IST 2019
'markdown_you' started at Sat May  4 11:21:46 IST 2019
'markdown_you' completed at Sat May  4 11:21:46 IST 2019
'markdown_marker' started at Sat May  4 11:21:46 IST 2019
Skipping marker...
'markdown_marker' completed at Sat May  4 11:21:46 IST 2019
'goword' started at Sat May  4 11:21:46 IST 2019
Skipping goword...
'goword' completed at Sat May  4 11:21:46 IST 2019
'gofmt' started at Sat May  4 11:21:46 IST 2019
'gofmt' completed at Sat May  4 11:21:48 IST 2019
'govet' started at Sat May  4 11:21:48 IST 2019
go: finding github.com/coreos/go-systemd v0.0.0-20180511133405-39ca1b05acc7
go: finding github.com/modern-go/reflect2 v1.0.1
go: finding github.com/mattn/go-colorable v0.0.9
go: finding github.com/olekukonko/tablewriter v0.0.0-20170122224234-a0225b3f23b5
go: finding github.com/dustin/go-humanize v0.0.0-20171111073723-bb3d318650d4
go: finding github.com/stretchr/testify v1.2.2
go: finding github.com/golang/protobuf v1.2.0
go: finding github.com/soheilhy/cmux v0.1.4
go: finding golang.org/x/crypto v0.0.0-20180608092829-8ac0e0d97ce4
go: finding go.uber.org/zap v1.9.1
go: finding github.com/pmezard/go-difflib v1.0.0
go: finding github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: finding github.com/inconshreveable/mousetrap v1.0.0
go: finding github.com/prometheus/procfs v0.0.0-20180612222113-7d6f385de8be
go: finding sigs.k8s.io/yaml v1.1.0     // this went through fine.
go: finding github.com/bgentry/speakeasy v0.1.0
go: finding github.com/google/uuid v1.0.0
go: finding gopkg.in/airbrake/gobrake.v2 v2.0.9
go: finding github.com/onsi/gomega v1.4.2
go: finding gopkg.in/cheggaaa/pb.v1 v1.0.25
go: finding github.com/grpc-ecosystem/grpc-gateway v1.4.1
go: finding github.com/gogo/protobuf v1.0.0
go: finding go.uber.org/multierr v1.1.0
go: finding github.com/golang/groupcache v0.0.0-20160516000752-02826c3e7903
go: finding github.com/json-iterator/go v1.1.5
go: finding gopkg.in/gemnasium/logrus-airbrake-hook.v2 v2.1.2
go: finding github.com/prometheus/client_model v0.0.0-20170216185247-6f3806018612
go: finding github.com/prometheus/client_golang v0.8.0
go: finding github.com/pkg/errors v0.8.0
go: finding github.com/xiang90/probing v0.0.0-20190116061207-43a291ad63a2
go: finding github.com/coreos/pkg v0.0.0-20160727233714-3ac0863d7acf
go: finding github.com/jonboulle/clockwork v0.1.0
go: finding github.com/kr/pty v1.0.0
go: finding github.com/onsi/ginkgo v1.6.0
go: finding github.com/google/btree v0.0.0-20180124185431-e89373fe6b4a
go: finding gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
go: finding go.uber.org/atomic v1.3.2
go: finding gopkg.in/yaml.v2 v2.2.2
```

I did not face the timeout issues which Travis reported. Not sure what is the cause here .

**Comment by spzala** [comment]
> > travis failed with "go: sigs.k8s.io/yaml@v1.1.0: unrecognized import path "sigs.k8s.io/yaml" (https fetch: Get https://sigs.k8s.io/yaml?go-get=1: net/http: TLS handshake timeout)".
> > it seems to have something to do with [42a7ea6](https://github.com/etcd-io/etcd/commit/42a7ea6d337a2fecc9162676407aa94e6a2b1f1e). /cc @rohitsardesai83
> 
> Hi @xiang90 , I cloned a fresh copy of etcd repo and tested with go 1.12.4
> 
> ```shell
> root@ubuntu-20:~/go/src/go.etcd.io/etcd# ./test
> Running with TEST_CPUS: 1,2,4
> Starting 'fmt' pass at Sat May  4 11:21:46 IST 2019
> 'shellcheck' started at Sat May  4 11:21:46 IST 2019
> 'shellcheck' completed at Sat May  4 11:21:46 IST 2019
> 'markdown_you' started at Sat May  4 11:21:46 IST 2019
> 'markdown_you' completed at Sat May  4 11:21:46 IST 2019
> 'markdown_marker' started at Sat May  4 11:21:46 IST 2019
> Skipping marker...
> 'markdown_marker' completed at Sat May  4 11:21:46 IST 2019
> 'goword' started at Sat May  4 11:21:46 IST 2019
> Skipping goword...
> 'goword' completed at Sat May  4 11:21:46 IST 2019
> 'gofmt' started at Sat May  4 11:21:46 IST 2019
> 'gofmt' completed at Sat May  4 11:21:48 IST 2019
> 'govet' started at Sat May  4 11:21:48 IST 2019
> go: finding github.com/coreos/go-systemd v0.0.0-20180511133405-39ca1b05acc7
> go: finding github.com/modern-go/reflect2 v1.0.1
> go: finding github.com/mattn/go-colorable v0.0.9
> go: finding github.com/olekukonko/tablewriter v0.0.0-20170122224234-a0225b3f23b5
> go: finding github.com/dustin/go-humanize v0.0.0-20171111073723-bb3d318650d4
> go: finding github.com/stretchr/testify v1.2.2
> go: finding github.com/golang/protobuf v1.2.0
> go: finding github.com/soheilhy/cmux v0.1.4
> go: finding golang.org/x/crypto v0.0.0-20180608092829-8ac0e0d97ce4
> go: finding go.uber.org/zap v1.9.1
> go: finding github.com/pmezard/go-difflib v1.0.0
> go: finding github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
> go: finding github.com/inconshreveable/mousetrap v1.0.0
> go: finding github.com/prometheus/procfs v0.0.0-20180612222113-7d6f385de8be
> go: finding sigs.k8s.io/yaml v1.1.0     // this went through fine.
> go: finding github.com/bgentry/speakeasy v0.1.0
> go: finding github.com/google/uuid v1.0.0
> go: finding gopkg.in/airbrake/gobrake.v2 v2.0.9
> go: finding github.com/onsi/gomega v1.4.2
> go: finding gopkg.in/cheggaaa/pb.v1 v1.0.25
> go: finding github.com/grpc-ecosystem/grpc-gateway v1.4.1
> go: finding github.com/gogo/protobuf v1.0.0
> go: finding go.uber.org/multierr v1.1.0
> go: finding github.com/golang/groupcache v0.0.0-20160516000752-02826c3e7903
> go: finding github.com/json-iterator/go v1.1.5
> go: finding gopkg.in/gemnasium/logrus-airbrake-hook.v2 v2.1.2
> go: finding github.com/prometheus/client_model v0.0.0-20170216185247-6f3806018612
> go: finding github.com/prometheus/client_golang v0.8.0
> go: finding github.com/pkg/errors v0.8.0
> go: finding github.com/xiang90/probing v0.0.0-20190116061207-43a291ad63a2
> go: finding github.com/coreos/pkg v0.0.0-20160727233714-3ac0863d7acf
> go: finding github.com/jonboulle/clockwork v0.1.0
> go: finding github.com/kr/pty v1.0.0
> go: finding github.com/onsi/ginkgo v1.6.0
> go: finding github.com/google/btree v0.0.0-20180124185431-e89373fe6b4a
> go: finding gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
> go: finding go.uber.org/atomic v1.3.2
> go: finding gopkg.in/yaml.v2 v2.2.2
> ```
> 
> I did not face the timeout issues which Travis reported. Not sure what is the cause here .

@rohitsardesai83 thanks for testing! And, probably not but I was wondering if it's related to go version as CI is not yet on 1.12.4 and the error seems across the board with new PRs (e.g. https://travis-ci.com/etcd-io/etcd/jobs/197642542#L630 is another PR)

**Comment by rohitsardesai50** [comment]
@spzala , in another instance , for  this PR #10693  , the CI passed : https://travis-ci.com/etcd-io/etcd/jobs/197674921. the dependency got downloaded correctly.

**Comment by rohitsardesai83** [comment]
@spzala , in another instance , for  this PR #10693  , the CI passed : https://travis-ci.com/etcd-io/etcd/jobs/197674921#L634. the dependency got downloaded correctly.

**Comment by jingyih** [comment]
I am new to go mod. Why do we need to download dependencies when we already have vendor folder?

**Comment by spzala** [comment]
> @spzala , in another instance , for this PR #10693 , the CI passed : https://travis-ci.com/etcd-io/etcd/jobs/197674921#L634. the dependency got downloaded correctly.

@rohitsardesai83 thanks, and thanks @jingyih for debugging it further!

**Comment by xiang90** [comment]
@spzala Can you try to retrigger the travis CI?

**Comment by jingyih** [comment]
Feels like maybe some of the go module flags are not configured properly. When building, dependencies code should be found in vendor folder, instead of downloading from a remote repo.

**Comment by spzala** [comment]
> @spzala Can you try to retrigger the travis CI?

@xiang90 sure, thanks, just restarted.

**Comment by spzala** [comment]
Go version is updated under https://github.com/etcd-io/etcd/pull/10766

**Comment by xiang90** [comment]
@spzala can you do a rebase to resolve the conflicts?

**Comment by spzala** [comment]
> @spzala can you do a rebase to resolve the conflicts?

@xiang90 yup, working on it. Thanks!

**Comment by spzala** [comment]
@xiang90 Hi, rebased so no conflicts now. The build failing seems related to the overall problems we are running into and not related to the changes. Thanks!

**Comment by gyuho** [comment]
@spzala Can we create an issue for those tests? Thanks!

**Comment by spzala** [comment]
> @spzala Can we create an issue for those tests? Thanks!

@gyuho I have added them to the umbrellas issue we have https://github.com/etcd-io/etcd/issues/10700 Also I noticed that, one of the failing test here `TestCtlV3AuthFromKeyPerm` is already added in the umbrella issue. Thanks!

**Comment by spzala** [comment]
Also, I am investing further for the `go.etcd.io/etcd/clientv3/balancer` unit test failure. Thanks!

**Comment by spzala** [comment]
@xiang90 @gyuho the build seems good now, the semaphorci failure is from known flaky tests that we have in the umbrella issue. Thanks!

File: proxy/grpcproxy/adapter/chan_stream.go [hunk 3]
```diff
@@ -120,10 +120,7 @@ func (s *chanStream) RecvMsg(m interface{}) error {
 		select {
 		case msg, ok := <-s.recvc:
 			if !ok {
-				if s.ctx.Err() != nil {
-					err := s.ctx.Err()
-					return err
-				}
+				return grpc.ErrClientConnClosing
```
Comment (xiang90) [inline proxy/grpcproxy/adapter/chan_stream.go]: why is this changed?

Comment (xiang90) [reply proxy/grpcproxy/adapter/chan_stream.go]: nvm

**Comment by xiang90** [comment]
@spzala can you squash the 10 commits into one or just a couple? the changes look good to me.

**Comment by spzala** [comment]
> @spzala can you squash the 10 commits into one or just a couple? the changes look good to me.

Thanks @xiang90 !! Just squashed into a single commit. Crossing fingers for build pass :-)

**Comment by spzala** [comment]
@xiang90 @gyuho thanks so much!
